overlord/udevmon.go

```diff
@@ -0,0 +1,124 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2018 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package overlord
+
+import (
+	"fmt"
+	"sync"
+
+	"github.com/snapcore/snapd/logger"
+
+	"github.com/pilebones/go-udev/crawler"
+	"github.com/pilebones/go-udev/netlink"
+)
+
+var (
+	udevMon *UDevMonitor
+	once    sync.Once
+)
+
+type udevMonitor interface {
+	Run() error
+	Stop()
+}
+
+type UDevMonitor struct {
+	netlinkConn *netlink.UEventConn
+	stop        chan struct{}
+	monitorStop chan struct{}
+	crawlerStop chan struct{}
+}
+
+func NewUDevMonitor() (*UDevMonitor, error) {
+	var err error
+	// There can only ever be one instance of netlink monitor per process,
+	// use singleton pattern to force this as otherwise it's very painful
+	// for the unit tests that do not delete overlord instances, or
+	// instantiate more than one overlord.
+	once.Do(func() {
+		udevMon = &UDevMonitor{
+			stop:        make(chan struct{}),
+			netlinkConn: &netlink.UEventConn{},
+		}
+		err = udevMon.netlinkConn.Connect()
+	})
+	if err != nil {
+		return nil, fmt.Errorf("failed to start uevent monitor: %s", err)
+	}
+	return udevMon, nil
+}
+
+func (m *UDevMonitor) Run() error {
```

Comment:
On initialization the hardware manager should subscribe to events and then enumerate devices (not the other way around). I don't know how this fits the overlord design yet but perhaps this should be done in the manager constructor.

The goroutine should be wrapped in a tomb and controlled by the manager. It should also go away once Stop() is called.

Reply:
> On initialization the hardware manager should subscribe to events and then enumerate devices (not the other way around). I don't know how this fits the overlord design yet but perhaps this should be done in the manager constructor.

Feels like monitoring and enumeration should be done in `Run()` rather than in `NewUdev*`, or am I missing something?

Reply:
That's right, done. Some of this is the the result of issues with tests and the fact that udev monitor is created via overlord multiple times in the tests and for some reason closing netlink socket and re-creating it fails with 'address already in use' (even though Close() is definately called on the socket). I'm still investigation / exploring this.